### DISCUSSION TITLE
How to set Agility SDK parameters?

### DISCUSSION THREAD

**Discussion by Marlax0** [terrafx/terrafx.interop.windows#349] [Q&A] [answered]
Edit: See my reply for the modern solution: https://github.com/terrafx/terrafx.interop.windows/discussions/349#discussioncomment-7681905

The [Getting Started with the Agility SDK](https://devblogs.microsoft.com/directx/gettingstarted-dx12agility/) guide describes how to set the parameters using two methods.

Unless I'm missing something, both these methods seem to be impossible with C#.
The guide states the symbols **must** be exported from the main executable for the process.
As of .NET 5 the executable is just a launcher that starts the runtime and then loads the application DLL.

Has anyone managed to figure this out?

Using `ID3D12SDKConfiguration.SetSDKVersion` works, but [only with Developer Mode enabled](https://learn.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12sdkconfiguration-setsdkversion#remarks).

There's also potentially `ID3D12SDKConfiguration1.CreateDeviceFactory` as detailed [here](https://microsoft.github.io/DirectX-Specs/d3d/IndependentDevices.html#api-details).
This doesn't require Developer Mode, but _getting_ or _querying for_ `ID3D12SDKConfiguration1` always fails for me.
Even if I call `ID3D12SDKConfiguration.SetSDKVersion` first it fails, so I dunno if it's a bug or the API just isn't usable yet.
> HRESULT: 80004002 (-2147467262)

I'll keep poking around, but any ideas would be appreciated.

**Comment by paulbartrum** [comment, marked as answer]
> Has anyone managed to figure this out?

I looked into this a while ago. I couldn't find a way to do this either, other than directly patching the EXE binary (yikes).

> getting or querying for ID3D12SDKConfiguration1 always fails for me.

Since this call happens before the Agility SDK is loaded, it requires a Windows update to function. According to the [DirectX discord](https://discord.gg/kkk2xWc) the independent devices functionality is only available on Windows Insider preview builds right now, but the DirectX team is working on rolling out a patch for retail versions of Win10/11. No timeframe was mentioned, so I guess we both just have to be patient :-)

**Reply by Marlax0** [reply]
Thank you for the heads up.
Yeah I also tried to patch the EXE, but after hours of failure I realized how in over my head I was and gave up.
At least with `ID3D12SDKConfiguration.SetSDKVersion` we can keep working with our fingers crossed.

It's good to know `ID3D12SDKConfiguration1` querying isn't a bug with TerraFX.
And since that was my only thread linking this question to TerraFX specifically, I'll go ahead and mark this as answered.

**Reply by Marlax0** [reply]
I come from the future.
Just wanted to note that D3D12 Independent Devices support has been released for Windows 10 at some point recently.
Not sure when, just happened to notice it today.

This should help anyone interested get started:
```csharp
const uint SDKVersion = 610; // Or whatever...

using ComPtr<ID3D12DeviceFactory> deviceFactory = null;

using ComPtr<ID3D12SDKConfiguration1> sdkConfiguration = null;

Guid guid = CLSID_D3D12SDKConfiguration;

D3D12GetInterface(&guid, UUID.Of<ID3D12SDKConfiguration1>(), (void**)sdkConfiguration.GetAddressOf());

byte[] sdkPathBytes = Encoding.ASCII.GetBytes(".\\D3D12\\"); // Or wherever...

fixed (byte* pSDKPathBytes = sdkPathBytes)
{
    sdkConfiguration.Get()->CreateDeviceFactory(SDKVersion, (sbyte*)pSDKPathBytes, __uuidof<ID3D12DeviceFactory>(), (void**)deviceFactory.GetAddressOf());
}
```

Then use `ID3D12SDKConfiguration1.CreateDevice` instead of `D3D12CreateDevice`.

**Reply by paulbartrum** [reply]
You should add a null terminator to your path string:
```c#
Encoding.ASCII.GetBytes(".\\D3D12\\\0");
```
I'm pretty sure it doesn't add the null terminator automatically (at least I couldn't find mention of it in the docs!)

**Reply by Marlax0** [reply]
At the very least it couldn't hurt.

**Reply by tannergooding** [reply]
Strings (both regular `"utf-16string"` and `"utf8-string"u8`) are already null-terminated.

Arbitrary spans may not be null-terminated (`"abc".AsSpan(0, 1)` is not null-terminated) and the null-terminator is not included in the reported length (`"".Length` is 0, but does have a null terminator still).
